### ISSUE TITLE
AssertionError on signal SIGQUIT or SIGINT using gevent worker

### ISSUE THREAD

**Issue by timonwong** [benoitc/gunicorn#1044]
Traceback

```
  File "/home/vagrant/.virtualenvs/gunicorn_test/lib/python2.7/site-packages/gunicorn/workers/base.py", line 153, in handle_quit
    time.sleep(0.1)
  File "/home/vagrant/.virtualenvs/gunicorn_test/lib/python2.7/site-packages/gevent/hub.py", line 82, in sleep
    hub.wait(loop.timer(seconds, ref=ref))
  File "/home/vagrant/.virtualenvs/gunicorn_test/lib/python2.7/site-packages/gevent/hub.py", line 348, in wait
    result = waiter.get()
  File "/home/vagrant/.virtualenvs/gunicorn_test/lib/python2.7/site-packages/gevent/hub.py", line 575, in get
    return self.hub.switch()
  File "/home/vagrant/.virtualenvs/gunicorn_test/lib/python2.7/site-packages/gevent/hub.py", line 337, in switch
    switch_out()
  File "/home/vagrant/.virtualenvs/gunicorn_test/lib/python2.7/site-packages/gevent/hub.py", line 341, in switch_out
    raise AssertionError('Impossible to call blocking function in the event loop callback')
```

It seems it's caused by 'time.sleep()'

Version: 19.3

**Comment by tilgovi** [comment]
I've seen this, too. It happens only during forced shutdown (INT, rather than QUIT/TERM). Only sometimes.

Is this what you see, too?

**Comment by reorx** [comment]
```
    def handle_quit(self, sig, frame):

        def _quit():
            self.alive = False
            time.sleep(0.1)
            sys.exit(0)

        if self.cfg.worker_class.__name__ == 'GeventWorker':
            import gevent
            print 'Using gevent.spawn to quit'
            gevent.spawn(_quit)
        else:
            _quit()
```

this is a quick dirty fix, after making it more clear I'll send a PR with more detail.

**Comment by benoitc** [comment]
It seems related to the use of `time.sleep` while closing a socket:
https://github.com/benoitc/gunicorn/blob/master/gunicorn/sock.py#L60

Applying #1088 should also fix it. Let me know.

**Comment by benoitc** [comment]
@timonwong bump

**Comment by parasyte** [comment]
@benoitc I just tried the patch, and it doesn't fix this issue with the gevent worker. I had to catch `AssertionError` separately here: https://github.com/benoitc/gunicorn/blob/79011d0c4a05353572a74e6ab743e4eb617c5d66/gunicorn/workers/base.py#L170 to quiet it.

**Comment by parasyte** [comment]
Reproduce with:

```python
# wsgi.py
import gevent.monkey
gevent.monkey.patch_all()

import bottle

app = bottle.Bottle()

@app.route("/ping")
def ping():
    """Test server availability."""
    return { 'message' : "pong" }
```

`$ gunicorn -b 0.0.0.0:8000 wsgi:app`

Press Ctrl+C to see the `AssertionError`

Using gunicorn 0.19.3

**Comment by benoitc** [comment]
what is the trace?

**Comment by parasyte** [comment]
@benoitc It's the same as above, though there are multiple assertions raised:

```
^CTraceback (most recent call last):
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/base.py", line 153, in handle_quit
[2015-09-01 15:04:16 -0700] [7274] [INFO] Handling signal: int
Traceback (most recent call last):
  File "gevent/core.pyx", line 360, in gevent.core.loop.handle_error (gevent/gevent.corecext.c:6139)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 467, in handle_error
    self.print_exception(context, type, value, tb)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 495, in print_exception
    traceback.print_exception(type, value, tb)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/traceback.py", line 125, in print_exception
    print_tb(tb, limit, file)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/traceback.py", line 69, in print_tb
    line = linecache.getline(filename, lineno, f.f_globals)
  File "/Users/jay/projects/pubnub-auth/env-auth/bin/../lib/python2.7/linecache.py", line 14, in getline
    lines = getlines(filename, module_globals)
  File "/Users/jay/projects/pubnub-auth/env-auth/bin/../lib/python2.7/linecache.py", line 40, in getlines
    return updatecache(filename, module_globals)
  File "/Users/jay/projects/pubnub-auth/env-auth/bin/../lib/python2.7/linecache.py", line 128, in updatecache
    lines = fp.readlines()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/base.py", line 153, in handle_quit
    time.sleep(0.1)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 117, in sleep
    hub.wait(loop.timer(seconds, ref=ref))
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 531, in wait
    result = waiter.get()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 774, in get
    return self.hub.switch()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 509, in switch
    switch_out()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 513, in switch_out
    raise AssertionError('Impossible to call blocking function in the event loop callback')
AssertionError: Impossible to call blocking function in the event loop callback
```

Then Ctrl+C a second time (the worker hasn't been killed yet):

```
^CTraceback (most recent call last):
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/base.py", line 153, in handle_quit
    time.sleep(0.1)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 117, in sleep
    hub.wait(loop.timer(seconds, ref=ref))
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 531, in wait
    result = waiter.get()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 774, in get
    return self.hub.switch()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 509, in switch
    switch_out()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 513, in switch_out
    raise AssertionError('Impossible to call blocking function in the event loop callback')
[2015-09-01 15:04:18 -0700] [7277] [ERROR] Exception in worker process:
Traceback (most recent call last):
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/arbiter.py", line 507, in spawn_worker
    worker.init_process()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/base.py", line 124, in init_process
    self.run()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 119, in run
    self.run_for_one(timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 81, in run_for_one
    self.wait(timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 35, in wait
    ret = select.select(self.sockets, [], self.PIPE, timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/select.py", line 78, in select
    result.event.wait(timeout=timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/event.py", line 86, in wait
    result = self.hub.switch()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 510, in switch
    return greenlet.switch(self)
AssertionError: Impossible to call blocking function in the event loop callback
Traceback (most recent call last):
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/arbiter.py", line 507, in spawn_worker
    worker.init_process()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/base.py", line 124, in init_process
    self.run()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 119, in run
    self.run_for_one(timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 81, in run_for_one
    self.wait(timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gunicorn/workers/sync.py", line 35, in wait
    ret = select.select(self.sockets, [], self.PIPE, timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/select.py", line 78, in select
    result.event.wait(timeout=timeout)
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/event.py", line 86, in wait
    result = self.hub.switch()
  File "/Users/jay/projects/pubnub-auth/env-auth/lib/python2.7/site-packages/gevent/hub.py", line 510, in switch
    return greenlet.switch(self)
AssertionError: Impossible to call blocking function in the event loop callback
```

**Comment by tilgovi** [comment]
Why do we call `time.sleep()` there?

**Comment by benoitc** [comment]
we don't now :) the log will say but i guess it was introduced to prevent a
race condition. which was wrong obviously.

- benoît

**Comment by parasyte** [comment]
Removing the sleep seems to kill the master before all of the workers have a chance to exit cleanly. Or in my tests, at least, not all of the workers print their exit message. This might be the same race condition.

**Comment by benoitc** [comment]
uh? the sleep is not related to the master at all.... tge socket is
actually closed in the worker.

**Comment by josegonzalez** [comment]
I'm definitely seeing this issue right now. I'm using `foreman` as my process manager and it appears that issuing `SIGINT` via `ctrl+c` doesn't stop gunicorn.I have to issue `SIGINT` twice in order to stop the worker.

This is on 19.3.0, so if it's been fixed in master, then we should release a new version :)

**Comment by josegonzalez** [comment]
Using master, it appears that I no longer need to send `SIGINT` twice, though I still get the AssertionError.

```
$ source .env && foreman start
02:32:55 web.1     | started with pid 58167
02:32:55 web_dev.1 | started with pid 58168
02:32:55 workers.1 | started with pid 58169
02:32:55 workers.1 | 2015-10-26 02:32:55,982 - INFO - Connection: haigha
02:32:55 workers.1 | 2015-10-26 02:32:55,985 - INFO - Trying to connect to host: localhost
02:32:55 workers.1 | 2015-10-26 02:32:55,995 - DEBUG - Loading consumer app.workers.convert:Consumer
02:32:56 workers.1 | [2015-10-26 06:32:56] app.workers.convert - INFO - Starting consumer app.workers.convert
02:32:56 web.1     | [2015-10-26 02:32:56 -0400] [58167] [INFO] Starting gunicorn 19.3.0
02:32:56 web.1     | [2015-10-26 02:32:56 -0400] [58167] [INFO] Listening at: http://0.0.0.0:5000 (58167)
02:32:56 web.1     | [2015-10-26 02:32:56 -0400] [58167] [INFO] Using worker: sync
02:32:56 web.1     | [2015-10-26 02:32:56 -0400] [58176] [INFO] Booting worker with pid: 58176
02:32:56 web_dev.1 |  * Running on http://0.0.0.0:5100/ (Press CTRL+C to quit)
02:32:56 web_dev.1 |  * Restarting with stat
02:32:56 web.1     | [2015-10-26 02:32:56 -0400] [58178] [INFO] Booting worker with pid: 58178
02:32:56 web.1     | [2015-10-26 02:32:56 -0400] [58179] [INFO] Booting worker with pid: 58179
02:32:56 web.1     | [2015-10-26 02:32:56 -0400] [58180] [INFO] Booting worker with pid: 58180
^CSIGINT received
02:32:57 system    | sending SIGTERM to all processes
SIGTERM received
02:32:57 web.1     | Traceback (most recent call last):
02:32:57 web_dev.1 | KeyboardInterrupt
02:32:57 workers.1 | KeyboardInterrupt
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gunicorn/workers/base.py", line 170, in handle_quit
02:32:57 web_dev.1 | Exception KeyError: KeyError(4421544848,) in <module 'threading' from '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.pyc'> ignored
02:32:57 workers.1 | Traceback (most recent call last):
02:32:57 web.1     | Traceback (most recent call last):
02:32:57 web_dev.1 | Exception SystemExit: 0 in <bound method Popen.__del__ of <subprocess.Popen object at 0x10fb69250>> ignored
02:32:57 workers.1 |   File "/Users/jose/.virtualenvs/seatdeck/bin/amqp-dispatcher", line 11, in <module>
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gunicorn/workers/base.py", line 170, in handle_quit
02:32:57 workers.1 |     sys.exit(main())
02:32:57 web.1     |     time.sleep(0.1)
02:32:57 web.1     | Traceback (most recent call last):
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 82, in sleep
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gunicorn/workers/base.py", line 170, in handle_quit
02:32:57 web.1     |     time.sleep(0.1)
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 82, in sleep
02:32:57 web.1     |     time.sleep(0.1)
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 82, in sleep
02:32:57 web.1     |     hub.wait(loop.timer(seconds, ref=ref))
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 348, in wait
02:32:57 web.1     |     result = waiter.get()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 575, in get
02:32:57 web.1     |     return self.hub.switch()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 337, in switch
02:32:57 web.1     | Traceback (most recent call last):
02:32:57 web.1     |     switch_out()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 341, in switch_out
02:32:57 web.1     |     raise AssertionError('Impossible to call blocking function in the event loop callback')
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gunicorn/workers/base.py", line 170, in handle_quit
02:32:57 web.1     | AssertionError: Impossible to call blocking function in the event loop callback
02:32:57 web.1     |     hub.wait(loop.timer(seconds, ref=ref))
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 348, in wait
02:32:57 web.1     |     hub.wait(loop.timer(seconds, ref=ref))
02:32:57 web.1     |     result = waiter.get()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 575, in get
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 348, in wait
02:32:57 web.1     |     return self.hub.switch()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 337, in switch
02:32:57 web.1     |     result = waiter.get()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 575, in get
02:32:57 web.1     |     switch_out()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 341, in switch_out
02:32:57 web.1     |     return self.hub.switch()
02:32:57 web.1     |     time.sleep(0.1)
02:32:57 web.1     |     raise AssertionError('Impossible to call blocking function in the event loop callback')
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 337, in switch
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 82, in sleep
02:32:57 web.1     |     switch_out()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 341, in switch_out
02:32:57 web.1     | AssertionError: Impossible to call blocking function in the event loop callback
02:32:57 web.1     |     raise AssertionError('Impossible to call blocking function in the event loop callback')
02:32:57 web.1     | AssertionError: Impossible to call blocking function in the event loop callback
02:32:57 web.1     |     hub.wait(loop.timer(seconds, ref=ref))
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 348, in wait
02:32:57 web.1     |     result = waiter.get()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 575, in get
02:32:57 web.1     |     return self.hub.switch()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 337, in switch
02:32:57 web.1     |     switch_out()
02:32:57 web.1     |   File "/Users/jose/.virtualenvs/seatdeck/lib/python2.7/site-packages/gevent/hub.py", line 341, in switch_out
02:32:57 web.1     |     raise AssertionError('Impossible to call blocking function in the event loop callback')
02:32:57 web.1     | [2015-10-26 02:32:57 -0400] [58167] [INFO] Handling signal: int
02:32:57 web.1     | AssertionError: Impossible to call blocking function in the event loop callback
02:32:57 workers.1 | terminated by SIGTERM
02:32:57 web_dev.1 | exited with code 0
02:32:57 web.1     | [2015-10-26 02:32:57 -0400] [58167] [INFO] Shutting down: Master
02:32:57 web.1     | exited with code 0
```

**Comment by jamadden** [comment]
Perhaps #1128 fixes this?

**Comment by josegonzalez** [comment]
@jamadden I can try out your pr later today :)

**Comment by tilgovi** [comment]
@jmadden merged that for you

**Comment by josegonzalez** [comment]
@jamadden @tilgovi seems that worked for me!

```
$ source .env && foreman start
02:28:34 web.1     | started with pid 82140
02:28:34 workers.1 | started with pid 82141
02:28:34 workers.1 | 2015-10-29 02:28:34,891 - INFO - Connection: haigha
02:28:34 workers.1 | 2015-10-29 02:28:34,893 - INFO - Trying to connect to host: localhost
02:28:34 workers.1 | 2015-10-29 02:28:34,900 - DEBUG - Loading consumer app.workers.convert:Consumer
02:28:35 web.1     | [2015-10-29 02:28:35 -0400] [82140] [INFO] Starting gunicorn 19.3.0
02:28:35 workers.1 | [2015-10-29 06:28:35] app.workers.convert - INFO - Starting consumer app.workers.convert
02:28:35 web.1     | [2015-10-29 02:28:35 -0400] [82140] [INFO] Listening at: http://0.0.0.0:5000 (82140)
02:28:35 web.1     | [2015-10-29 02:28:35 -0400] [82140] [INFO] Using worker: sync
02:28:35 web.1     | [2015-10-29 02:28:35 -0400] [82146] [INFO] Booting worker with pid: 82146
02:28:35 web.1     | [2015-10-29 02:28:35 -0400] [82147] [INFO] Booting worker with pid: 82147
02:28:35 web.1     | [2015-10-29 02:28:35 -0400] [82148] [INFO] Booting worker with pid: 82148
02:28:35 web.1     | [2015-10-29 02:28:35 -0400] [82152] [INFO] Booting worker with pid: 82152
^CSIGINT received
02:28:36 system    | sending SIGTERM to all processes
02:28:36 workers.1 | terminated by SIGTERM
$ ps aux|grep python
jose            82160   0.0  0.0  2433796    684 s008  S+    2:28AM   0:00.00 grep python
```

**Comment by tilgovi** [comment]
\o/ thanks @jmadden

**Comment by timonwong** [comment]
@tilgovi 
I can confirm that it works after applying @jamadden's patch.
Sorry for my looooong delay, and thanks again :)

**Comment by tilgovi** [comment]
Closed by #1128
